src/core/xref.js

```diff
@@ -76,11 +91,70 @@ function createXrefQuery(xrefs) {
     },
     new Set()
   );
-  return { keys: [...queryKeys].map(JSON.parse) };
+  return [...queryKeys].map(JSON.parse);
+}
+
+// adds data to cache
+async function cacheResults(data, cache) {
+  await cache.ready;
+  const promises = Object.entries(data).map(([key, value]) =>
+    cache.put(key.toLowerCase(), value)
+  );
+  await cache.put("__CACHE_TIME__", new Date());
+  await Promise.all(promises);
+}
+
+/**
+ * looks for keys in cache and resolves them
+ * @param {Array} keys query keys
+ * @param {IDBCache} cache
+ * @returns {Object}
+ *  @property {Object} found resolved data from cache
+ *  @property {Array} notFound keys not found in cache
+ */
+async function resolveFromCache(keys, cache) {
+  await cache.ready;
+
+  const cacheTime = await cache.match("__CACHE_TIME__");
+  const bustCache = cacheTime && new Date() - cacheTime > CACHE_MAX_AGE;
+  if (bustCache) {
+    await cache.clear();
+    return { found: {}, notFound: keys };
+  }
+
+  const promises = keys.map(({ term }) => cache.match(term.toLowerCase()));
+  const cachedData = await Promise.all(promises);
+  return keys.reduce(resolve, { found: {}, notFound: [] });
+
+  function resolve(collector, key, i) {
```

Comment:
nit: maybe call this something else, as it could be confused with promise `resolve`